pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,513 @@
+// ProcessCluster loops through implemented deprovisioners
+// nolint:gocyclo
+func (c *Controller) ProcessCluster(ctx context.Context) (Result, error) {
+	// range over the different deprovisioning methods. We'll only let one method perform an action
+	for _, d := range []deprovisioner{
+		// Emptiness and Consolidation are mutually exclusive, so this loop only executes one deprovisioner
+		c.emptiness,
+		c.consolidation,
+	} {
+		var result Result
+		var err error
+		// Capture new cluster state before each attempt
+		candidates, err := c.candidateNodes(ctx, d.shouldNotBeDeprovisioned)
+		if err != nil {
+			return ResultFailed, fmt.Errorf("determining candidate nodes, %w", err)
+		}
+		// If there are no candidate nodes, move to the next deprovisioner
+		if len(candidates) == 0 {
+			continue
+		}
+		// Sort candidates by each deprovisioner's priority, and deprovision the first one possible
+		candidates = d.sortCandidates(candidates)
+		result, err = c.executeDeprovisioning(ctx, d, candidates...)
+		if err != nil {
+			logging.FromContext(ctx).Errorf("deprovisioning nodes, %s", err)
+			continue
+		}
+		if result == ResultRetry {
+			return ResultRetry, nil
+		}
+		// Return if success, continue to other nodes otherwise.
+		if result == ResultSuccess {
+			return result, nil
+		}
+		// Continue to next deprovisioner if nothing was done
+		if result == ResultNothingToDo {
+			continue
+		}
+		return result, nil
+	}
+	// All deprovisioners did nothing, so return nothing to do
+	return ResultNothingToDo, nil
+}
+
+func (c *Controller) executeDeprovisioning(ctx context.Context, d deprovisioner, nodes ...candidateNode) (Result, error) {
+	// Given candidate nodes, compute best deprovisioning action
+	attempts := 0
+	var cmd deprovisioningCommand
+	var err error
+	success := false
+	// Each attempt will try at least one node, limit to that many attempts.
+	for attempts < len(nodes) {
+		cmd, err = d.computeCommand(ctx, attempts, nodes...)
+		if err != nil {
+			return ResultFailed, err
+		}
+		// Validate the command with a TTL
+		ok, err := c.validateCommand(ctx, cmd, d)
+		if err != nil {
+			return ResultFailed, fmt.Errorf("validating command, %w", err)
+		}
+		// If validation succeeds, execute the command. Otherwise, we need to compute another command and
+		// try the rest of the candidate nodes. If any of the remaining candidateNodes are still valid
+		// for an action, continue with deprovisioning, even though we might not be taking the most optimal decision.
+		// This is meant to stop an issue if another controller in the cluster coincidentally creates churn at the same interval.
+		if ok {
+			success = true
+			break
+		}
+		attempts++
+	}
+	// Validation for all candidateNodes failed, so nothing can be done for this deprovisioner.
+	if !success {
+		return ResultNothingToDo, nil
+	}
+	// Execute the command
+	result, err := c.executeCommand(ctx, cmd)
+	if err != nil {
+		return ResultFailed, err
```

Comment:
I agree, this is redundant here, and goes with your comment above with collapsing the result into a `reconcile.Result`, but with the current results, I'd rather return a redundant but accurately named result, rather than a innacurately named result.